pkg/auth/iam/iam.go

```diff
@@ -63,19 +63,17 @@ type ErrorResponse struct {
 // NewFilter creates new Filter instance
 func NewFilter(client iam.Client) *Filter {
 	options := &FilterInitializationOptions{}
-	if s, exists := os.LookupEnv(constant.EnvSubdomainValidationEnabled); exists {
-		subdomainValidationEnabled, err := strconv.ParseBool(s)
-		if err != nil {
-			logrus.Errorf("Parse %s env error: %v", constant.EnvSubdomainValidationEnabled, err)
-		}
-		if subdomainValidationEnabled == true {
-			options.AllowSubdomainMatchRefererHeaderValidation = true
-		}
-	}
 	return &Filter{iamClient: client, options: options}
 }
 
 // NewFilterWithOptions creates new Filter instance with Options
+// Example:
+//
+//	iam.NewFilterWithOptions(iamClient, &FilterInitializationOptions{
+//		AllowSubdomainMatchRefererHeaderValidation: true
+//		SubdomainValidationEnabled: true,
+//		SubdomainValidationExcludedNamespaces: ["foundations"]
+//	})
 func NewFilterWithOptions(client iam.Client, options *FilterInitializationOptions) *Filter {
```

Comment:
I have a question: does that mean all the golang services need to be modified to use this method while the foundations merge back to core? or do the foundations only need to change some specific endpoints?

Reply:
yes all golang services need to be modified to use the new init filter option.

Reply:
if this is the case, I suggest not creating the new one. Otherwise, we need to make the code change to the service. In the future, the foundations will be part of the core.

Reply:
i still don't understand.
what do you mean by not creating the new one?

Reply:
are you suggesting that we modify the `NewFilter` function instead of using the `NewFilterWithOptions`?
I need to clarify that we need the `FilterInitializationOptions` to handle the appropriate cases between justice/foundations by the values from environment variables.

Reply:
Yes, from my view, no need to create another NewFilterWithOptions, reason:
1. from your description, the opt-in/out are controlled by the envs, so no matter if it's in justice or foundations, it can initialize as if the envs provided. I suppose it will be disabled by default.
2. no need to change all the services. Once foundations merge back to core, that means we only have 1 copy of the code; if we have NewFilterWithOptions, that means before the filter initializing, we need to identify whether it's a multi-tenant mode or not, why not always use NewFilter and let the envs to control the behavior.

Reply:
I still have some concerns:
1. the `NewFilterWithOptions` has been there for a while, I'm not sure if it is only used in foundations, or somewhere else, I will confirm this later.
2. we still eventually need to change all the services to add the environment variables
3. in my opinion, although the code becomes a little more verbose, passing the env variable values via `FilterInitializationOptions` will make the spec clearer.
4. it will prevent this plugin from being tightly coupled with the details of environment variables. instead, the spec is explicitly stated in the `FilterInitializationOptions`.

Reply:
for 1, I originally thought it was a new method, and yes, we need to leave it there
for 2. Adding the environment and code changes are 2 different things
for 3, but I do not think so. This depends on how we encapsulate them and how to design the plugin. Again, if all the services support multi-tenant, that means we have to use FilterInitializationOptions, and why need the NewFilter, it's only for back-compatible.
for 4, yes, this is a good perspective. But if there's no encapsulation, that means all the services need to deal with the environments and the relationship between environment variables by themselves. like subdomainValidationEnabled:true -> AllowSubdomainMatchRefererHeaderValidation: true

Reply:
alright.
I got your point, but I still need your suggestion to proceed from here.
we DO need (almost) all services to support multi-tenant.

should we:
1. keep the `NewFilterWithOptions` and the `FilterInitializationOptions`
2. modify the `NewFilter` to read the environment variables and set `FilterInitializationOptions` implicitly,
3. all services still use `NewFilter`

or should we:
1. keep the `NewFilterWithOptions` and the `FilterInitializationOptions`
2. create a new module on each service to handle multi-tenant requirements, then use it to map the environment variable to `FilterInitializationOptions`
3. all services use `NewFilterWithOptions`

Reply:
because each service will have multi-tenant, I suggest:
1. keep the NewFilterWithOptions and the FilterInitializationOptions
2. have a FilterInitializationOptions builder to deal with the multi-tenant
3. all services use NewFilterWithOptions

Reply:
I have added a function to build FilterInitializationOptions based on the env variable.